src/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs

```diff
@@ -0,0 +1,125 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Threading
+{
+    internal static partial class WaitSubsystem
+    {
+        [EagerStaticClassConstruction] // used during lazy class construction
+        private static class HandleManager
+        {
+            private const int InitialCapacity = 4;
+
+            /// <summary>
+            /// Prevents reusing a deleted index until at least there are this many deleted indexes in the free list. The
+            /// purpose is to allow free handles to remain free for some time so that we can discover invalid attempts at using
+            /// free handles.
+            /// </summary>
+            private const int MinimumFreeIndexCountBeforeReuse = 64;
+
+            private static WaitableObject[] s_objects = new WaitableObject[InitialCapacity];
+            private static int s_count;
+            private static LowLevelQueue<int> s_freeIndexes;
```

Comment:
Having free list that is parallel with `WaitableObject[]` may be better.  A lot less code than Queue, not out of memory on attempt to free the code, easier to detect bad handles.

- When the handle index is free, it would be index of next free element.
- When the handle index is occupied, it would be "cookie" to detect invalid handles. Something like:

```
struct Handle
{
    public WaitableObject Value;
    public int Link;
}

static Handle[] s_handles = ...


public static WaitableObject FromHandle(IntPtr handle)
{
     int index = ((int)handle - 1) & (s_handles.Length - 1);
     ref Handle handle = s_handles[index];
     
     WaitableObject value = handle.Value

     if (value == null || handle.Link != (int)handle)
          throw InvalidOperationException.NewInvalidHandle(); 

     return value;
}
```